Can't build on linux

Hello,

first of, thanks so much for releasing this plugin as open-source, it looks really cool! I found a few issues when compiling this on linux though, seems like the code is using some of the Visual Studio templates (like dot) and there are some undeclared variables here and there; I don't have much c++ competence though, so I added a stack trace of what I'm getting out of cmake:

[TwistSpline_cmake_log.txt](https://github.com/blurstudio/TwistSpline/files/2397126/TwistSpline_cmake_log.txt)

A few errors that I tracked down already are:

- twistSplineNode.cpp
  - line 33 `#include <maya/MFNPluginData.h>` is duplicate and also has a capital N letter, remove the line
- riderConstraint.h
  - `riderConstraint::` on MObject assignment, apparently this only works in Visual Studio, removing that will fix one issue
- mpkdtree.h
  - `dot` seems to be a microsoft's custom template for Visual Studio, a workaround to make the build pass that issue was to copy the custom `dot` template from twistSpline.h to mpkdtree.h and commenting it out in twistSpline.h, so it won't be declared twice
- twistSpline.h
  - line 117,  `&v` variable looks like it's not assigned but it's called from that line, I couldn't understand what was going on with that last function so I stopped here

I hope this helps tracking down build issues!

Cheers!
Matteo

Hi again!

Been trying to get this working in Maya 2020 on Linux, but I'm getting some errors on compiling again. I can't really figure out what's going on here, If I compile for Maya 2022 or Maya 2019 the build is fine, but not on 2020 - I tried every version of the Devkit without success.
Are you able to compile this on 2020 for Windows? If yes then I'll flag this on Autodesk's forum or file a bug if necessary.

Here's the output of the error:
```make
Scanning dependencies of target TwistSpline
[ 33%] Building CXX object CMakeFiles/TwistSpline.dir/twistSplineNode.cpp.o
[ 33%] Building CXX object CMakeFiles/TwistSpline.dir/riderConstraint.cpp.o
[ 50%] Building CXX object CMakeFiles/TwistSpline.dir/twistSplineData.cpp.o
[ 66%] Building CXX object CMakeFiles/TwistSpline.dir/pluginMain.cpp.o
[ 83%] Building CXX object CMakeFiles/TwistSpline.dir/twistTangentNode.cpp.o
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:41,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/pluginMain.cpp:25:
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:367:29: error: 'Autodesk::Maya::OpenMaya20200000::MHWRender::MSelectionInfo::~MSelectionInfo()' is private within this context
  367 | class OPENMAYARENDER_EXPORT MDrawContext : public MFrameContext
      |                             ^~~~~~~~~~~~
In file included from /sww/packages/maya/2020/payload/include/maya/MFrameContext.h:20,
                 from /sww/packages/maya/2020/payload/include/maya/MViewport2Renderer.h:32,
                 from /sww/packages/maya/2020/payload/include/maya/MPxDrawOverride.h:27,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:39,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/pluginMain.cpp:25:
/sww/packages/maya/2020/payload/include/maya/MSelectionContext.h:143:2: note: declared private here
  143 |  ~MSelectionInfo();
      |  ^
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:41,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/pluginMain.cpp:25:
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:367:29: error: 'Autodesk::Maya::OpenMaya20200000::MHWRender::MPassContext::~MPassContext()' is private within this context
  367 | class OPENMAYARENDER_EXPORT MDrawContext : public MFrameContext
      |                             ^~~~~~~~~~~~
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:322:2: note: declared private here
  322 |  ~MPassContext();
      |  ^
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:41,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.cpp:50:
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:367:29: error: 'Autodesk::Maya::OpenMaya20200000::MHWRender::MSelectionInfo::~MSelectionInfo()' is private within this context
  367 | class OPENMAYARENDER_EXPORT MDrawContext : public MFrameContext
      |                             ^~~~~~~~~~~~
In file included from /sww/packages/maya/2020/payload/include/maya/MFrameContext.h:20,
                 from /sww/packages/maya/2020/payload/include/maya/MViewport2Renderer.h:32,
                 from /sww/packages/maya/2020/payload/include/maya/MPxGeometryOverride.h:26,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.cpp:42:
/sww/packages/maya/2020/payload/include/maya/MSelectionContext.h:143:2: note: declared private here
  143 |  ~MSelectionInfo();
      |  ^
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:41,
                 from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.cpp:50:
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:367:29: error: 'Autodesk::Maya::OpenMaya20200000::MHWRender::MPassContext::~MPassContext()' is private within this context
  367 | class OPENMAYARENDER_EXPORT MDrawContext : public MFrameContext
      |                             ^~~~~~~~~~~~
/sww/packages/maya/2020/payload/include/maya/MDrawContext.h:322:2: note: declared private here
  322 |  ~MPassContext();
      |  ^
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/pluginMain.cpp:25:
/home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h: In member function 'virtual void TwistSplineDrawOverride::handleTraceMessage(const Autodesk::Maya::OpenMaya20200000::MString&) const':
/home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:147:35: warning: format not a string literal and no format arguments [-Wformat-security]
  147 |   fprintf(stderr, message.asChar());
      |                                   ^
In file included from /home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.cpp:50:
/home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h: In member function 'virtual void TwistSplineDrawOverride::handleTraceMessage(const Autodesk::Maya::OpenMaya20200000::MString&) const':
/home/mdilena/dev/vendor/twist-spline-bazel/TwistSpline/twistSplineNode.h:147:35: warning: format not a string literal and no format arguments [-Wformat-security]
  147 |   fprintf(stderr, message.asChar());
      |                                   ^
make[2]: *** [CMakeFiles/TwistSpline.dir/build.make:63: CMakeFiles/TwistSpline.dir/pluginMain.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/TwistSpline.dir/build.make:89: CMakeFiles/TwistSpline.dir/twistSplineNode.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/TwistSpline.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -14,6 +14,7 @@ set(SOURCE_FILES
     "twistSpline_maya.h"
     "twistSplineData.h"
     "twistSplineNode.h"
+    "twistSplineUtils.h"
     "riderConstraint.h"
     "decomposeRotation.h"
     "twistTangentNode.h"

--- a/mpkdtree.h
+++ b/mpkdtree.h
@@ -27,6 +27,8 @@ SOFTWARE.
 #include <tuple>
 #include <algorithm>
 #include <math.h>
+#include "twistSplineUtils.h"
+
 
 struct ltGetter {
     size_t splitDim;

--- a/twistSpline.h
+++ b/twistSpline.h
@@ -30,61 +30,7 @@ SOFTWARE.
 #include <cmath>
 #include <numeric>
 #include "mpkdtree.h"
-
-
-// To make my choice of abstraction work, I have to write wrappers for the template abstractions
-template <typename PointArray>
-inline void resize(PointArray &a, unsigned size) {
-	a.resize(size);
-}
-
-template <typename PointArray>
-inline unsigned size(const PointArray &a) {
-	return a.size();
-}
-
-template <typename Vector, typename Float = double>
-inline Float dot(const Vector &a, const Vector &b) {
-	return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
-}
-
-template <typename Vector, typename Float = double>
-inline Float length(const Vector &a) {
-	return sqrt(dot<Vector, Float>(a, a));
-}
-
-template <typename Vector, typename Float = double>
-inline Vector normalized(const Vector &a) {
-	return a / length<Vector, Float>(a);
-}
-
-template <typename Vector>
-inline Vector cross(const Vector &a, const Vector &b) {
-	return Vector((a[1] * b[2] - a[2] * b[1]), (a[0] * b[2] - a[2] * b[0]), (a[0] * b[1] - a[1] * b[0]));
-}
-
-/// Like projection except you get the perpendicular component of the vector
-template <typename Vector, typename Float=double>
-inline Vector reject(const Vector &onto, const Vector &n) {
-	return normalized<Vector, Float>(n - ((dot<Vector, Float>(n, onto) / dot<Vector, Float>(onto, onto)) * onto));
-}
-
-template <typename Quat>
-inline Quat qmult(const Quat &q, const Quat &r){
-    return Quat(r[0]*q[0]-r[1]*q[1]-r[2]*q[2]-r[3]*q[3],
-            r[0]*q[1]+r[1]*q[0]-r[2]*q[3]+r[3]*q[2],
-            r[0]*q[2]+r[1]*q[3]+r[2]*q[0]-r[3]*q[1],
-            r[0]*q[3]-r[1]*q[2]+r[2]*q[1]+r[3]*q[0]);
-}
-
-template <typename Vector, typename Quat>
-inline Vector rotateBy(const Vector &vec, const Quat &quat){
-	Quat r(0.0, vec[0], vec[1], vec[2]);
-	Quat c(quat[0], -quat[1], -quat[2], -quat[3]);
-	Quat ret = qmult(qmult(quat, r), c);
-	return Vector(ret[1], ret[2], ret[3]);
-}
-
+#include "twistSplineUtils.h"
 
 
 /**

--- a/twistSplineUtils.h
+++ b/twistSplineUtils.h
@@ -1,0 +1,81 @@
+/*
+MIT License
+
+Copyright (c) 2018 Blur Studio
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#pragma once
+
+
+// To make my choice of abstraction work, I have to write wrappers for the template abstractions
+template <typename PointArray>
+inline void resize(PointArray &a, unsigned size) {
+	a.resize(size);
+}
+
+template <typename PointArray>
+inline unsigned size(const PointArray &a) {
+	return a.size();
+}
+
+template <typename Vector, typename Float = double>
+inline Float dot(const Vector &a, const Vector &b) {
+	return a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
+}
+
+template <typename Vector, typename Float = double>
+inline Float length(const Vector &a) {
+	return sqrt(dot<Vector, Float>(a, a));
+}
+
+template <typename Vector, typename Float = double>
+inline Vector normalized(const Vector &a) {
+	return a / length<Vector, Float>(a);
+}
+
+template <typename Vector>
+inline Vector cross(const Vector &a, const Vector &b) {
+	return Vector((a[1] * b[2] - a[2] * b[1]), (a[0] * b[2] - a[2] * b[0]), (a[0] * b[1] - a[1] * b[0]));
+}
+
+/// Like projection except you get the perpendicular component of the vector
+template <typename Vector, typename Float=double>
+inline Vector reject(const Vector &onto, const Vector &n) {
+	return normalized<Vector, Float>(n - ((dot<Vector, Float>(n, onto) / dot<Vector, Float>(onto, onto)) * onto));
+}
+
+template <typename Quat>
+inline Quat qmult(const Quat &q, const Quat &r){
+    return Quat(r[0]*q[0]-r[1]*q[1]-r[2]*q[2]-r[3]*q[3],
+            r[0]*q[1]+r[1]*q[0]-r[2]*q[3]+r[3]*q[2],
+            r[0]*q[2]+r[1]*q[3]+r[2]*q[0]-r[3]*q[1],
+            r[0]*q[3]-r[1]*q[2]+r[2]*q[1]+r[3]*q[0]);
+}
+
+template <typename Vector, typename Quat>
+inline Vector rotateBy(const Vector &vec, const Quat &quat){
+	Quat r(0.0, vec[0], vec[1], vec[2]);
+	Quat c(quat[0], -quat[1], -quat[2], -quat[3]);
+	Quat ret = qmult(qmult(quat, r), c);
+	return Vector(ret[1], ret[2], ret[3]);
+}
+
+

